c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -240,7 +241,9 @@ class DeviceCachingAllocator {
       // Attempt allocate
       || alloc_block(params, false)
       // Free all non-split cached blocks and retry alloc.
-      || (free_cached_blocks() && alloc_block(params, true));
+      || (free_cached_blocks() && alloc_block(params, true))
+      // Run expensive free memory callbacks and try again
+      || (trigger_expensive_free_memory_callbacks(params) && get_free_block(params));
```

Comment:
So the big problem with the naive implementation, as you've done here, is that we're missing the reasoning about what happens if the PyTorch GC *triggers* an allocation and we attempt to reenter the allocation. I'm pretty sure this can be made to work, but it needs to be explicitly reasoned about.